lib/langchain/agent/function_call_agent/function_call_agent.rb

```diff
@@ -0,0 +1,40 @@
+# frozen_string_literal: true
+
+module Langchain::Agent
+  class FunctionCallAgent < Base
+    attr_reader :conversation, :tools
+
+    #
+    # Initializes the Agent
+    #
+    # @param llm [Object] The LLM client to use
+    # @param tools [Array<Object>] The tools that can be invoked
+    #
+    def initialize(llm:, tools:)
+      Langchain::Tool::Base.validate_tools!(tools: tools)
+
+      @conversation = Langchain::Conversation.new(llm: llm)
+      @tools = tools
+    end
+
+    #
+    # Ask a question and responds or call the appropriate function
+    #
+    # @param question [String] Question to ask the LLM
+    # @return [String] Answer to the question
+    #
+    def run(question:)
+      conversation.set_functions(tools.map(&:tool_schema))
+      ai_response = conversation.message(question)
+
+      if (function = ai_response.additional_kwargs[:function_call])
+        tool = tools.find { |tool| tool.tool_name == function["name"] }
+        tool.execute(**JSON.parse(function["arguments"]).transform_keys(&:to_sym)).tap do |result|
+          conversation.add_function_call_result(function["name"], result)
```

Comment:
should we call the llm one more time to get a more friendly answer to the user?

Reply:
what do you think to leave that to the tool? If the tool wants to use an llm to get the answer then it can. I dont know if I like the idea of all responses to be through the llm (making an extra request, adding cost / delay)

Reply:
I get your point, but I think it's expected that agents do this 🤔 for example it is done in the `SQLQueryAgent`